From 40bbb161e72fb609608d53b9d64c56bb961a6ee2 Mon Sep 17 00:00:00 2001
From: Kyle Nekritz <knekritz@fb.com>
Date: Mon, 25 Feb 2019 17:07:20 -0800
Subject: [PATCH] Avoid arithmetic operation on uint16 read from the wire.

Summary:
This could overflow previously.

CVE-2019-3560

Reviewed By: yfeldblum

Differential Revision: D14152362

fbshipit-source-id: c0ebb3fc59b49c7c23e6bcb90458c19cd891be65
---
 fizz/record/PlaintextRecordLayer.cpp     |  4 +---
 fizz/record/test/PlaintextRecordTest.cpp | 10 ++++++++++
 2 files changed, 11 insertions(+), 3 deletions(-)

diff --git a/fizz/record/PlaintextRecordLayer.cpp b/fizz/record/PlaintextRecordLayer.cpp
index e33ef9ee..958c412a 100644
--- a/fizz/record/PlaintextRecordLayer.cpp
+++ b/fizz/record/PlaintextRecordLayer.cpp
@@ -39,9 +39,7 @@ folly::Optional<TLSMessage> PlaintextReadRecordLayer::read(
         if (buf.chainLength() < (cursor - buf.front()) + length) {
           return folly::none;
         }
-        length +=
-            sizeof(ContentType) + sizeof(ProtocolVersion) + sizeof(uint16_t);
-        buf.trimStart(length);
+        buf.trimStart(static_cast<size_t>(kPlaintextHeaderSize) + length);
         continue;
       } else if (msg.type != ContentType::change_cipher_spec) {
         skipEncryptedRecords_ = false;
diff --git a/fizz/record/test/PlaintextRecordTest.cpp b/fizz/record/test/PlaintextRecordTest.cpp
index 21842288..1f65e718 100644
--- a/fizz/record/test/PlaintextRecordTest.cpp
+++ b/fizz/record/test/PlaintextRecordTest.cpp
@@ -115,6 +115,16 @@ TEST_F(PlaintextRecordTest, TestSkipAndWait) {
   EXPECT_TRUE(queue_.empty());
 }
 
+TEST_F(PlaintextRecordTest, TestSkipOversizedRecord) {
+  read_.setSkipEncryptedRecords(true);
+  addToQueue("170301fffb");
+  auto longBuf = IOBuf::create(0xfffb);
+  longBuf->append(0xfffb);
+  queue_.append(std::move(longBuf));
+  EXPECT_FALSE(read_.read(queue_).hasValue());
+  EXPECT_TRUE(queue_.empty());
+}
+
 TEST_F(PlaintextRecordTest, TestWaitBeforeSkip) {
   read_.setSkipEncryptedRecords(true);
   addToQueue("170301000501234567");
